### ISSUE TITLE
SystemJS - proof of concept with Create React App ejected Webpack - unable to access module properties

### ISSUE THREAD

**Issue by shanplourde** [systemjs/systemjs#2328]
- SystemJS Version: 
- Which library are you using?
  - [ X] system.js
  - [ ] s.js
  - [ ] system-node.cjs
- Which extras are you using?
  - [ ] AMD extra
  - [ ] Named Exports
  - [ ] Named Register
  - [ ] Transform
  - [ ] Use Default
  - [ ] Global
  - [ ] Dynamic Import Maps
- Are you using any [custom hooks](/docs/hooks.md)?

### Question

Hi, I have a proof of concept where I am trying to use a Create React App ejected Webpack 5 config as the basis for a module that is exported to System JS format. The config has been updated to support System JS format. The consumer of the module is not able to access the properties that the module exposes. I have followed the System JS guidelines for Webpack config and Webpack dev server config. When I try to load the module, System.import() seems to load the module's main entry, but I am not able to access any of the module's properties. Is there something else I would need to do? 

Consumer app index.html:
```html
        <script src="https://cdn.jsdelivr.net/npm/systemjs@6.8.3/dist/system.min.js"></script>
        <script type="systemjs-importmap">
            {
                "imports": {
                    "wp5-ms": "//localhost:3000/static/js/main.chunk.js"
                }
            }
        </script>
```

Consumer app TypeScript:

```javascript
    React.useEffect(() => {
        const loadExample = async () => {
            const wp5Example = await global.System.import('wp5-ms');
            console.log('wp5Example', wp5Example);
        };

        loadExample();
    }, []);
```

Module main entry file:

```javascript
import "systemjs-webpack-interop/auto-public-path";

console.log("hi outside");
const foo = { foo: "bar" };
export default foo;
```

Console output in consumer application:

![image](https://user-images.githubusercontent.com/1335741/116962545-3e84b380-ac74-11eb-85c9-a54176efb456.png)

Here is the high level output of the SystemJS module that is generated with my Webpack config:

![image](https://user-images.githubusercontent.com/1335741/116962651-93282e80-ac74-11eb-9795-50d5df1ba8f2.png)

**Comment by shanplourde** [comment]
Also, the "hi outside" is not logged to console.

**Comment by shanplourde** [comment]
I think I am finding that the "optimization" section of Webpack is causing issues. Notably when chunking is enabled. Is Webpack chunking supported with SystemJS? 

I am confused about what may or may not function with respect to Webpack's chunking because the documentation sometimes seems to indicate that it is supported. For example, single-spa suggests that chunking is supported:

> Consider turning off webpack hashing for both entry and code split bundles. It is often easier to add in a commit hash during deployment of your microfrontend via your CI environment variables


But then the single-spa documentation regarding CRA https://single-spa.js.org/docs/faq/#create-react-app says:

> Remove Webpack optimizations block, because they add multiple webpack chunks that don't load each other

SystemJS documentation about Webpack compatibility https://github.com/systemjs/systemjs#compatibility-with-webpack doesn't seem to indicate that chunking is not possible.

**Comment by shanplourde** [comment]
I noted https://github.com/systemjs/systemjs/issues/1920, which indicates that rollup code splitting would be compatible. But based on the above notes in the current SystemJS and Single Spa documentation, I am curious to know if there is some sort of Webpack approach today that would work for this purpose.

**Comment by joeldenning** [comment]
Webpack bundles that use the `optimization` config for code splits do not properly expose their exports. See https://single-spa.js.org/docs/recommended-setup#build-tools-webpack--rollup for a list of changes to your webpack config that will likely get it loading properly with systemjs. The main things are to check that the entry point isn't an array (unless your main entry is the last item in the array, in which case it's okay), along with checking the libraryTarget and the optimization property.

**Comment by shanplourde** [comment]
Hi @joeldenning  Is the Webpack limitation by design or a bug if you know? 

I was able to get my integration working by disabling code splitting. That being said, are there any code splitting strategies within a single SystemJS exported module that you know of that would reliably work with SystemJS?

**Comment by joeldenning** [comment]
Using webpack dynamic import is an effective way of code splitting when using libraryTarget system.

The main limitation with webpack's `optimization` config is that it generates a bunch of files that must be loaded in a specific order in order for the bundle to correctly instantiate. The majority of the webpack ecosystem uses html-webpack-plugin which inserts `<script>` elements to load all the separate bundles in the correct order.

SystemJS modules are each a single javascript file, that can have dependencies on each other. Ideally, the main output bundle would correctly list its dependencies in a way that SystemJS could understand and then load. However, webpack does not do this, as webpack code splits generated via `optimization` are in webpackJsonp format rather than System.register format, and also do not know what their dependencies are. This limitation is something that the author of webpack was open to being solved, but just said would take a lot of work. See https://github.com/webpack/webpack/issues/8833#issuecomment-468181544 where this was discussed further.

With the limitation in mind, some people have tried to make it work anyway by doing `Promise.all(['/main.js', '/chunk1.js', 'chunk2.js', 'vendors.js'])`, but that is problematic because it does not guarantee order of execution.

**Comment by naivefun** [comment]
@shanplourde have you made it work?

**Comment by shanplourde** [comment]
@naivefun No, I didn't get it to work in my proof of concept. Chunking is not supported with the Webpack system library target. I guess other library formats would be needed with Webpack if chunking were desired, or perhaps create additional Webpack outputs of type system as an alternate strategy to chunking.
